gemini/sample-apps/finance-advisor-spanner/database.py

```diff
@@ -0,0 +1,230 @@
+"""This file is for database operations done by the application """
+
+# pylint: disable=line-too-long, invalid-name, use-dict-literal, use-list-literal
+import os
+
+from dotenv import load_dotenv
+from google.api_core.client_options import ClientOptions
+from google.cloud import spanner
+import pandas as pd
+import streamlit as st
+from streamlit_extras.stylable_container import stylable_container
+
+load_dotenv()
+
+instance_id = os.getenv("instance_id")
+database_id = os.getenv("database_id")
+api_endpoint = os.getenv("api_endpoint")
+
+options = ClientOptions(api_endpoint=api_endpoint)
+spanner_client = spanner.Client(client_options=options)
+
+instance = spanner_client.instance(instance_id)
+database = instance.database(database_id)
+
+
+def spanner_read_data(query: str) -> pd.DataFrame:
+    """This function helps read data from spanner"""
+    with database.snapshot() as snapshot:
+        results = snapshot.execute_sql(query)
+        rows = list()
+        for row in results:
+            rows.append(row)
+        cols = [x.name for x in results.fields]
+        result_df = pd.DataFrame(rows, columns=cols)
+    return result_df
+
+
+def spanner_read_data_list(query: str) -> list:
+    """This function helps read data from spanner and returns data as a list"""
+    with database.snapshot() as snapshot:
+        results = snapshot.execute_sql(query)
+        rows = list()
+        for row in results:
+            rows.append(row)
+    return rows
+
+
+def spanner_read_data_withparameters(query: str, vector_input: list) -> pd.DataFrame:
+    """This function helps read data from spanner with input params"""
+    with database.snapshot() as snapshot:
+        results = snapshot.execute_sql(
+            query,
+            params={"vector": vector_input},
+        )
+        rows = list()
+        for row in results:
+            rows.append(row)
+        cols = [x.name for x in results.fields]
+        result_df = pd.DataFrame(rows, columns=cols)
+
+    return result_df
```

Comment:
```suggestion
def spanner_read_data(query: str, params: dict = None) -> pd.DataFrame:
    """This function helps read data from Spanner"""
    with database.snapshot() as snapshot:
        results = snapshot.execute_sql(query, params=params)
        rows = list(results)
        cols = [x.name for x in results.fields]
        return pd.DataFrame(rows, columns=cols)
```

Do you really need three different methods for this? Seems like a lot of duplication. I think it could be condensed into one method with slightly different calls.

Reply:
fixed